WIP Breadcrumbs and you - getting to know each other

JMuff22 opened the pull request:
Trying to fix breadcrumb links in certain page. The broken links are caused by a non-existent url in the breadcrumbs. An easy fix to this is to create an `index.md` file inside every subfolder in `docs/`. 

As an example in `docs/computing/quantum-computing/helmi/first-quantum-job/` the breadcrumbs to `Quantum computing` and `Helmi` are broken. These can be fixed by adding in an `index.md` file in `docs/computing/quantum-computing/`. In this case there in an `overview.md` that should be displayed so the fix can simply be to symbolically link `index.md` to `overview.md` with `ln -s overview.md index.md`. 

The main concern with this solution is that some links to docs.csc.fi will break. 


The page can be previewed here:
https://csc-guide-preview.rahtiapp.fi/origin/fix-breadcrumb-links/

JMuff22 commented:
`breadcrumbs.html` now correctly renders the breadcrumbs for top level `index.html` files. 

This works by taking all pages where a page doesn't have any ancestors (i.e index.html/md files). A Jinja variable is set which takes the `page.url` and derives the breadcrumb from that. 

Macro from: https://shubhamjain.co/2018/09/17/capitalizing-first-letter-in-jinja/

JMuff22 commented:
Consequence bug with this method: When you click the breadcrumb, up 1 level to the newly fixed links the navigation bar on the left side resets to 0.

rkronberg commented:
We could try creating a dummy index file instead of symlinks to other (e.g. overview.md) files. I'll test this and commit if it fixes the consequence bug.

rkronberg commented:
> We could try creating a dummy index file instead of symlinks to other (e.g. overview.md) files. I'll test this and commit if it fixes the consequence bug.

Does not seem to work without appropriately reorganizing files. So our options are to hack something poorly maintainable together or then harmonize the folder hierarchy throughout docs and commit to fixing all links that will break in the process.

Opinions @joonas-somero @Nortamo @attesillanpaa ?

JMuff22 commented:
Issue pointed out by @joonas-somero.

*Computing > Systems > Available systems* the link Computing gets `href="//"` because `/computing/available-systems/` has fewer levels in breadcrumbs.

https://csc-guide-preview.rahtiapp.fi/origin/fix-breadcrumb-links/computing/available-systems/

JMuff22 commented:
Upon further review, to fully fix breadcrumbs, we need to add top level `index.md` files to all directories. Symbolic links work but close the contents bar on the left when clicked in a breadcrumb. Therefore all `overview.md` files and other top level descriptor files have been moved to be called `index.md`. The `index.md` file must also be present in the contents tree in `mkdocs.yml`. 

This is a substantial change because it affects all links to the previous field as well as all links from outside docs.csc.fi. 

A solution can be seen in the branch `fix-breadcrumb-links-2`: https://github.com/CSCfi/csc-user-guide/tree/fix-breadcrumb-links-2 with ~ 60 files changed. 

`breadcrumbs.html` still needs fixing to point to the correct link and where e.g cloud-->pouta-->security guidelines for pouta works fine but click the pouta link then the "Pouta" breadcrumb goes 1 level up not the cloud breadcrumb.

JMuff22 commented:
Merged fix-breadcrumb-links-2 into fix-breadcrumbs-links. This:
- Adds a test bash script which can be used with `bash tests/check_index.sh`. I added an entry into `.travis.yml` but not sure if it works yet. There are still unresolved directories: `docs/support/tutorials` which will be fixed in the future. 
- From the test I added in a few missing `index.md` files. 

To-do:
- Move support/ up 1 level and fix mkdocs.yml --> instead of `support/*` there should be top level directories `docs/tutorials`, `docs/wn` 
- remove patch-fixes in `breadcrumbs.html`

JMuff22 commented:
Added a plugins.yml file which contains all the plugins. This is inherited by `mkdocs.yml` and this is done to clena up `mkdocs.yml`. [8613c8f](https://github.com/CSCfi/csc-user-guide/pull/1480/commits/8613c8f9d35ec059291638387225c45e54063880)

With the `mkdocs-redirects` plugin now all links that pointed to `support/tutorials`, `support/faq` and `support/wn` should continue to work regardless of if we update the links or not. With this in place this PR can be merged without worrying about any outside links breaking.

JMuff22 commented:
To-do (30-01-23): 
- `csc-overrides/stylesheets/sidebars.css` changing how "Section", "Page" or "SectionPage" is highlighted with/without mouse cursor hover.
- What's new entry
- Investigate broken(?) breadcrumbs on `tutorials/gis` and `tutorials/env-guide`. 
- Docs preview app updated with new packages and working
- Message to docs developers letting them know of the changes
    - Index file in new folders
    - New python packages 
    - New Changes to folder structure
    - Breadcrumbs preview addition (`breadcrumbs-debug.html`) 
    - 


To view the PR locally create a new conda env:
```bash
conda env create -f docs/tutorials/conda/conda-docs-env-1.2.yaml
conda activate docs-env
```
Or in your current env `pip install mkdocs-section-index` `pip install mkdocs-section-index`. OR via docker...


Things to be reviewed:
- `404.md` message
- What's new entry (to-do)

joonas-somero commented:
Preview is now back in business!

rkronberg commented on docs/apps/krona.md:
```diff
@@ -57,7 +57,7 @@ ktImportTaxonomy -q 2 -t 3 KrakenOutput -o KronaReport.html
 
-Alternatively you can copy these files to a publicly accessible data bucket in Allas, and study the results with your local web browser. You can use commands [a-flip](../../data/Allas/using_allas/a_commands/#a-flip) or [a-publish](../data/Allas/using_allas/a_commands.md#a\
+Alternatively you can copy these files to a publicly accessible data bucket in Allas, and study the results with your local web browser. You can use commands [a-flip](../../data/Allas/using_allas/a_commands/#a-flip) or [a-publish](../data/Allas/using_allas/a_commands/#a\
 -publish) to do the copying:
```
I think the a-publish link is broken

rkronberg reviewed (commented):
Breadcrumbs in Allas sections seem to be still broken. E.g. https://csc-guide-preview.rahtiapp.fi/origin/fix-breadcrumb-links/data/Allas/using_allas/a_commands/

rkronberg commented on docs/apps/lammps.md:
```diff
@@ -35,11 +35,11 @@ LAMMPS on Puhti or Mahti for optimal performance.
 Consult these pages on [how to create batch jobs on
-Puhti](../computing/running/creating-job-scripts-puhti.md) and [on
-Mahti](../computing/running/creating-job-scripts-mahti.md).
+Puhti](../computing/running/creating-job-scripts-puhti/) and [on
+Mahti](../computing/running/creating-job-scripts-mahti/).
```
These links are broken

joonas-somero commented:
> Breadcrumbs in Allas sections seem to be still broken. E.g. https://csc-guide-preview.rahtiapp.fi/origin/fix-breadcrumb-links/data/Allas/using_allas/a_commands/

It just so happens that a solution to this might already be in a PR: https://github.com/CSCfi/csc-user-guide/pull/1554

rkronberg commented on docs/apps/rapids.md:
```diff
@@ -81,14 +81,14 @@ list-packages
 applications. See our [general instructions on how to take the fast local
 storage into
-use](../computing/running/creating-job-scripts-puhti.md#local-storage).
+use](../computing/running/creating-job-scripts-puhti/#local-storage).
```
link broken

joonas-somero replied:
Another case of a link split on two lines. I'll try to come up with a way to find all of these.

rkronberg replied:
Thanks! I think this would be valuable since I fear it is quite common practice to split links.. I mean why not, it is valid markdown syntax and helps to keep the content readable in a text editor.

joonas-somero replied:
I think there are around 200 of them.

joonas-somero replied:
Looks like the "checkbuild" test catches two-line links if they are to a `.md` file that doesn't exist. This is because MkDocs throws a warning and "checkbuild" doesn't allow that.

joonas-somero replied:
148cac0 fixes the link in question.

rkronberg commented on docs/cloud/pouta/index.md:
```diff
@@ -29,4 +29,4 @@
 * [Pouta videos](pouta-videos.md)
 * [Known problems and limitations](known-problems.md)
-* [Pouta FAQ](../../support/faq/index.md#pouta)
+* [Pouta FAQ](../../faq#pouta)
```
link broken

rkronberg commented on docs/computing/allas.md:
```diff
@@ -51,19 +51,19 @@ The _allas_ module provides several tools that you can use to move data between
 The Allas documentation includes two tutorials that are especially designed for Puhti and Mahti users:
 
-* [Examples for using Allas in CSC supercomputers](../data/Allas/allas-examples.md)
+* [Examples for using Allas in CSC supercomputers](../../data/Allas/allas-examples)
```
link broken (trailing slash needed)

joonas-somero commented:
> > Breadcrumbs in Allas sections seem to be still broken. E.g. https://csc-guide-preview.rahtiapp.fi/origin/fix-breadcrumb-links/data/Allas/using_allas/a_commands/
> 
> It just so happens that a solution to this might already be in a PR: #1554

As was commented in `mkdocs.yml` of #1554, `/data/Allas/using_allas/directory_object_error/` isn't included, so it doesn't have any breadcrumbs.

rkronberg commented on docs/data/sensitive-data/sd-desktop-audited.md:
```diff
@@ -361,4 +360,4 @@ If you are still waiting for a decision about your amendment application when th
 ## Troubleshooting
 
-Find solutions to common problems in the [Troubleshooting paragraph](../../data/sensitive-data/sd_desktop.md#troubleshooting) under SD Desktop.
+Find solutions to common problems in the [Troubleshooting paragraph](#troubleshooting) under SD Desktop.
```
This link is probably wrong, the original one seems correct.

rkronberg commented on docs/support/index.md:
```diff
@@ -1,6 +1,6 @@
 # Supporting materials
 
-## [Tutorials](tutorials/index.md)
```
Should tutorials and FAQ be moved away from this index since they are no longer in the support directory?

joonas-somero replied:
I think they fit on there quite well. The whole page seems to be there just in case someone types the url from memory/has a bookmark.

joonas-somero replied:
#509

rkronberg commented on docs/support/training-material.md:
```diff
@@ -1,14 +1,14 @@
 
-Get the basics right! In addition the the [Tutorials](tutorials/index.md)
-and [FAQ](faq/index.md) collections, this page links to a selection of
+Get the basics right! In addition the the [Tutorials](../../tutorials)
```
Links broken

joonas-somero commented:
A simplified solution to the problem was formed in #1573. This PR will now serve as ~a monument~ documentation for later PRs. Thanks for the hard work, everyone!
